### DISCUSSION TITLE
データの消去について

### DISCUSSION THREAD

**Discussion by silli-3072** [lit-iphone-course/lit-iphone-course#16] [Question-original] [answered]
お疲れ様です。Realmのデータ消去について質問です。
テーブルビューのスワイプアクションでdeleteが選択された時にデータの消去を実行したいです。
記事（https://swift.hiros-dot.net/?p=645#toc8）を参考に実装してみたのですが、

`let results = realm.objects(ToDo.self).filter()`

filter()の中に何を記述していいかわかりません。記事の事以外にもdeleteを選択した行を取得してその部分に記述しようとしましたがこちらもうまく実装できませんでした。（そもそもdelete選択時の取得方法に確信を持てませんでした。）
どのように実装すればいいか教えていただきたいです。
GitHubコチラになります→https://github.com/silli-3072/LA4_ToDoApp

**Comment by appleEmily** [comment, marked as answer]
<img width="524" alt="image" src="https://user-images.githubusercontent.com/49332380/188772226-42ccc6e8-7f1e-4f97-ac17-5b42866fd222.png">
やりたいことは、この画像のようなものであってるよね？

やることとしては、
①スワイプしてdeleteされるところを検知するコードを書く。(参考記事)
②realmから選択されている行の要素を消す。選択されている行はtableViewなのでIndexpathで取得できるよ〜
③tableをreloadする

参考になる記事
https://taracodelog.com/uitableview-edit/
https://qiita.com/TsukasaHasegawa/items/4dfaa87add7b6866f858
https://taracodelog.com/edit-realm/

**Reply by silli-3072** [reply]
やりたいことは添付いただいた写真のようなことで、deleteを押されたときにレルムのデータを消去するような実装をしたいです！
質問時の認識として

`let results = realm.objects(ToDo.self).filter()`

この部分のカッコ内に正しいコードを書ければ実装できると思っていました。カッコ内に`indexPath.row`と記述するとエラーが出てしまいます。（添付画像）
<img width="1512" alt="スクリーンショット 2022-09-07 18 12 22" src="https://user-images.githubusercontent.com/105117616/188840418-49364315-0757-4916-97ea-c301668df93e.png">

改めて、主に二つ目の記事を参考にコードを変えてみましたが、うまくいきませんでした💦GitHub（https://github.com/silli-3072/LA4_ToDoApp）最新にプッシュ済みです。

**Reply by txxxxc** [reply]
データ取得方法としては、データを取得したい時にその都度realmから全件取得するのではなく画面が読み込まれた時に取得して、結果を保持するのが一般的です！
なので今回の例で言うとuserDataに値が入っていると思うのでそこからindexPath.rowの値を作って要素にアクセスと良さそう！！！

**Reply by silli-3072** [reply]
リロードの部分でエラーが出てしまい想像通りの実装とはなっていませんが消去部分はうまく実装できました！！
ありがとうございました！

**Reply by txxxxc** [reply]
リロードの部分も試してみて難しそうだったら全然答えるよー！

**Reply by silli-3072** [reply]
お疲れ様です！リロード上手くできておらず、教えていただきたいです。
 ```
override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        tableView.reloadData()
        
    }

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func tableView(_ tableView: UITableView, trailingSwipeActionsConfigurationForRowAt indexPath: IndexPath) -> UISwipeActionsConfiguration? {
        let userData = realm.objects(ToDo.self)
        
        let deleteAction = UIContextualAction(style: .destructive, title: "Delete") { [self] (action, view, completionHandler) in
        
            let item = userData[indexPath.row]
            
            try! realm.write {
                self.realm.delete(item)
            }
        
            completionHandler(true)
        }
        
        return UISwipeActionsConfiguration(actions: [deleteAction])
        
        tableView.reloadData()
    }
```
viewWillAppearとスワイプアクションの実装の二箇所の部分でtableView.reloadData()を記述しているのですがviewWillAppearではエラーがスワイプアクションではシュミレーターで確認できなくて、どこを変えればレルムの消去をテーブルビューに反映できるでしょうか？

[参考にしたサイト](https://qiita.com/TsukasaHasegawa/items/4dfaa87add7b6866f858) ・[GitHub](https://github.com/silli-3072/LA4_ToDoApp) こちらです！

**Reply by txxxxc** [reply]
結論から言うと、スワイプアクションの部分で`reloadData`がうまく動いていない原因はreturn後にreloadDataを呼び出しているからです！関数はreturnするとそれ以降の処理が実行されず終わってしまうのでreturnの後には処理を書かないほうが良いです！

あと、deleteActionという変数する際に書かれている処理(`let item = userData[indexPath.row]`から`completionHandler(true)`の所)がcellを削除した時に行われる処理なので、その中にtableviewのリロードは必要になります！
それを踏まえて一回試してみてくださいー！

実装方法ちょっと違うけど、いい感じの記事見つけたから貼っておくね↓
https://teratail.com/questions/290989

**Reply by silli-3072** [reply]
delete選択時のテーブルビュー更新はできました！
ただ、viewWillAppearの方でリロードができず、新規追加した際にアプリのタスクを切らないと表示されません。
上記のコードのままなのですがそちらのエラーはどう対応すれば良いか教えていただきたいです

**Reply by txxxxc** [reply]
多分Realmに保存したあとにtableViewをリロードしてないのが原因だと思う！
追加画面から戻った時にtableViewをリロードさせる方法は以下のコメントを参考にしつつ実装してもらえると🙏↓
https://github.com/lit-iphone-course/lit-iphone-course/discussions/14#discussioncomment-3565671

**Reply by silli-3072** [reply]
めっちゃできました！！！変数名が違う事によるエラーでした...
